window-manager/src/main/java/com/codelab/foldables/window_manager/MainActivity.kt

```diff
@@ -20,59 +20,63 @@ package com.codelab.foldables.window_manager
 import android.view.View
 import androidx.appcompat.app.AppCompatActivity
 import androidx.constraintlayout.widget.ConstraintSet
-import androidx.core.util.Consumer
-import androidx.window.WindowLayoutInfo
-import androidx.window.WindowManager
+import androidx.window.layout.FoldingFeature
+import androidx.window.layout.WindowInfoRepository
+import androidx.window.layout.WindowInfoRepository.Companion.windowInfoRepository
+import androidx.window.layout.WindowLayoutInfo
+import androidx.window.layout.WindowMetricsCalculator
 import com.codelab.foldables.window_manager.databinding.ActivityMainBinding
-import java.util.concurrent.Executor
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
 
 class MainActivity : AppCompatActivity() {
-
-    private lateinit var wm: WindowManager
-    private val layoutStateChangeCallback = LayoutStateChangeCallback()
     private lateinit var binding: ActivityMainBinding
 
-    private fun runOnUiThreadExecutor(): Executor {
-        val handler = Handler(Looper.getMainLooper())
-        return Executor() {
-            handler.post(it)
-        }
-    }
+    private lateinit var windowInfoRepository: WindowInfoRepository
+    private val scope = MainScope()
 
+    @ExperimentalCoroutinesApi
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         binding = ActivityMainBinding.inflate(layoutInflater)
         setContentView(binding.root)
 
-        wm = WindowManager(this)
-    }
+        windowInfoRepository = windowInfoRepository()
 
-    override fun onAttachedToWindow() {
-        super.onAttachedToWindow()
-        wm.registerLayoutChangeCallback(
-            runOnUiThreadExecutor(),
-            layoutStateChangeCallback
-        )
+        obtainWindowMetrics()
+        onWindowLayoutInfoChange(windowInfoRepository)
     }
 
-    override fun onDetachedFromWindow() {
-        super.onDetachedFromWindow()
-        wm.unregisterLayoutChangeCallback(layoutStateChangeCallback)
+    override fun onDestroy() {
+        super.onDestroy()
+        scope.cancel()
     }
 
-    private fun printLayoutStateChange(newLayoutInfo: WindowLayoutInfo) {
+    private fun obtainWindowMetrics() {
+        val wmc = WindowMetricsCalculator.getOrCreate()
         binding.windowMetrics.text =
-            "CurrentWindowMetrics: ${wm.currentWindowMetrics.bounds.flattenToString()}\n" +
-                "MaximumWindowMetrics: ${wm.maximumWindowMetrics.bounds.flattenToString()}"
+            "CurrentWindowMetrics: ${wmc.computeCurrentWindowMetrics(this).bounds.flattenToString()}\n" +
+                "MaximumWindowMetrics: ${wmc.computeMaximumWindowMetrics(this).bounds.flattenToString()}"
+    }
 
+    private fun onWindowLayoutInfoChange(windowInfoRepository: WindowInfoRepository) {
+        scope.launch {
```

Comment:
@pfmaggi for simplicity and to avoid adding a new dependency to the code and sample (and make it more complex) I've used the main scope. If you think we should use `lifecycleScope`, please let me know.  We probably would have to add a note about it as well in the Codelab-content.

Reply:
I usually prefer to use `lifecycleScope` to pause/restart the collection when the app is in background. Up to you if you think that this will not introduce issues.

Reply:
I definitely think that your suggestion is the right approach when creating apps and using arch components, but wondering if asking for an alpha version of an extra library may add some concerns when using WM (e.g. devs can think: "I need an extra component on an alpha stage?"). I haven't seen any issue with this approach, and in the codelab content (document) we also reference and highlight this alternative of using lifecycleScope, so I think we can leave it as it is.
The important thing is that we mention both alternatives so learners know both options :-) Thoughts?

Reply:
Sounds good as both approach work for this use case (probably for all use cases). I agree that we should go with the easier option.

Thanks

Reply:
Awesome! thanks for letting me know :-)